src/liballoc/collections/btree/search.rs

```diff
@@ -46,6 +46,15 @@ where
     }
 }
 
+/// Returns the index in the node at which an equivalent of the key exists
+/// or could exist, and whether it exists in the node itself. If it doesn't
+/// exist in the node, it may still exist in the subtree with that index
+/// (unless the node is a leaf), or it may be the right index to insert at.
+///
+/// This function is defined for every BorrowType (immutable, mutable, owned),
+/// but it never mutates or takes ownership of anything itself, nor does it
+/// return any indication of BorrowType, so it doesn't need a mutable or owned
+/// version of `keys`.
```

Comment:
I have to admit that the second paragraph here is not terribly useful for me. Is this saying something the caller should be aware of (but there is no `keys` parameter), or explaining something about the implementation?

(Also looks like GH stopped sending email notifications on force-pushes, so I entirely missed that you added this. :/ )

Reply:
It's not something callers should be aware of and I agree it's not terribly useful. I just don't know how or where else to "explain that all these searches must be done in the immutable version to avoid calling `into_key_slice_mut` on the shared root".

Reply:
Where in the implementation is this fact that it is using immutable slices reflected? IMO that's where the comment should go.

Reply:
So you mean, move the paragraph a few lines down, and I guess with double instead of triple slashes?

Reply:
So the key line is `node.keys()`, right? Yes, that's the place. I wouldn't just move the comment though, also edit it a bit. For example:

> Crucially, we use `keys()` here, i.e., we work with immutable data. We must not call `keys_mut()` as that does not work for the shared root, and `node` may well be the shared root here. Using `keys()` is fine here even if `BorrowType` is mutable, as all we return is an index -- not a reference.